src/zcash/Note.hpp

```diff
@@ -150,20 +213,20 @@ class SaplingNotePlaintext : public BaseNotePlaintext {
 
     template <typename Stream, typename Operation>
     inline void SerializationOp(Stream& s, Operation ser_action) {
-        unsigned char leadingByte = 0x01;
-        READWRITE(leadingByte);
-
-        if (leadingByte != 0x01) {
-            throw std::ios_base::failure("lead byte of SaplingNotePlaintext is not recognized");
-        }
```

Comment:
If every code path that deserializes a Sapling note plaintext immediately calls `plaintext_version_is_valid`, then this change is fine. But that seems fragile to me; removing the serialization error from here means that all future note plaintext versions will be parsed in the same format as the currently-defined versions, which I don't think is intended. Even though the note plaintext is fixed-length, the delineation of fields within it might change, and/or a future transaction version might make the ciphertext longer.

I think we should continue to throw an error here if we do not recognise the leading byte (i.e. `if (leadByte > 0x02)` as it is unsigned), and then further restrict the valid `leadByte` values contextually in `plaintext_version_is_valid`. This matches how we handle transaction versions.

Reply:
I agree, we should restore this exception. Let's make sure it's restored and the logic that _caught_ exceptions is also restored in `Note.cpp`.

Reply:
restored!

Reply:
Oh, I made a comment requesting the opposite. I don't think it's helpful to have duplicative checks, which just complicate reasoning about which error is actually reported. I actually would prefer that we remove *all* exception throwing from *all* serialisation.

Reply:
In another comment I pointed out that preserving these exception throws is a conservative option for us to avoid blocking this PR.